nck/readers/mytarget_reader.py

```diff
@@ -0,0 +1,252 @@
+# along with this program; if not, write to the Free Software Foundation,
+# Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+import click
+import nck.helpers.api_client_helper as api_client_helper
+from nck.clients.api_client import ApiClient
+from nck.commands.command import processor
+from nck.readers.reader import Reader
+from nck.streams.json_stream import JSONStream
+from nck.utils.args import extract_args
+import requests
+from typing import Dict
+from nck.utils.secret_retriever import get_secret
+from botocore.config import Config
+import logging
+import json
+import boto3
+import itertools
+
+
+@click.command(name="read_mytarget")
+@click.option("--mytarget-client-id", required=True)
+@click.option("--mytarget-client-secret", required=True)
+@click.option("--mytarget-mail", required=True)
+@click.option("--mytarget-agency", required=True)
+@processor("mytarget-client-id", "mytarget-client-secret")
+def mytarget(**kwargs):
+    return MyTargetReader(**extract_args("mytarget_", kwargs))
+
+
+secret_keys = json.loads(get_secret(
+    'arn:aws:secretsmanager:eu-west-3:127773427021:secret:access-key-sanofi-ANrWKh',
+    'eu-west-3')
+)
+
+
+s3_resource = boto3.resource(
+    's3',
+    aws_access_key_id=secret_keys.get('ACCESS_KEY'),
+    aws_secret_access_key=secret_keys.get('SECRET_KEY'),
+    region_name='eu-west-3',
+    config=Config(
+        retries={
+            'max_attempts': 10
+        }
+    )
+)
+
+
+class MyTargetReader(Reader):
+
+    def __init__(
+        self,
+        client_id,
+        client_secret,
+        mail,
+        agency,
+        **kwargs
+    ):
+        self.client_id = client_id
+        self.client_secret = client_secret
+        self.mail = mail
+        self.agency = agency
+        self.access_token = None
+        self.agency_client_token = None
+
+    def create_agency_token_request(self, agency_client):
+        url = 'https://target.my.com/api/v2/oauth2/token.json'
+        headers = {
+            'Content-Type': 'application/x-www-form-urlencoded',
+            'Host': 'target.my.com'
+        }
+        payload = dict(
+            grant_type='agency_client_credentials',
+            client_id=self.client_id,
+            client_secret=self.client_secret,
+            agency_client_name=agency_client
+        )
+        return dict(
+            url=url,
+            headers=headers,
+            data=payload
+        )
+
+    def retrieve_agency_client_response(self):
+        return requests.post(**self.create_agency_token_request(self.agency)).json()
+
+    def is_agency_client_token_new(self, rsp):
+        return "error" not in rsp.keys()
+
+    def __write_agency_client_token_s3(self, content: Dict[str, str]):
+        s3_resource.Object('token-mytarget', f'{self.agency_client_token}_last_token.json').put(Body=json.dumps(content))
+        logging.info('The token has been uploaded to the bucket.')
```

Comment:
This code is very specific to your use case. What if a developer who is working in a GCP environment wants to use the MyTarget reader? Why do we store the token? Is it hard to get a new one?

Reply:
True, going to remove this part and proceed differently.